evalml/utils/gen_utils.py

```diff
@@ -378,3 +378,13 @@ def _not_nan(pd_data):
 
     mask = np.logical_and(_not_nan(pd_data_1), _not_nan(pd_data_2))
     return pd_data_1.iloc[mask], pd_data_2.iloc[mask]
+
+
+def woodwork_wrapper(func):
```

Comment:
I think this would also work for pipelines/estimators but I haven't tested it. I think another alternative is to have this be a function rather than a wrapper.

So we would go from 
```python
@woodwork_wrapper
def my_method(self, X, y, ...):
```
to 
```python
def my_method(self, X, y, ...):
    X, y = woodwork_wrapper(X, y)
```

I think this would be the most general (since it can apply to more than just methods that take `self`) and also a bit less mysterious. For example, in `HighlyNullDataCheck`, our docstrings say we accept pd DataFrames, but then we call 

```python
    X = _convert_woodwork_types_wrapper(X.to_dataframe())
```

and `to_dataframe()` is not a dataframe method. So someone reading the code wight be confused as to how it works. 

I think this implementation is great (and it potentially works for more than just data checks) but I wanted to propose an alternative!

Reply:
I agree here, I think if users were to look at our source code, they might be confused why we can make the `to_dataframe()` or `to_series()` calls.

Reply:
@freddyaboulton and @bchen1116 thanks for this!

When @angela97lin and I discussed this recently, we liked the decorator approach. However your opinions here are making me think there's not much difference between doing this in a decorator and doing it in a helper method, except that the decorator adds some complexity by hiding the conversion. In light of that, I'd be in favor of using a helper over a decorator. @angela97lin what do you think?

Reply:
Thanks for the suggestions! I agree, even when implementing the wrapper, it seemed a bit silly :d I like the idea of not using the wrapper.

In response to all of this: is an additional helper method really necessary? Or it is it okay to just continue with adding
`X = _convert_to_woodwork_structure(X)` and `y = _convert_to_woodwork_structure(y)` to methods that need it? 

@freddyaboulton I've thought about a helper that takes in both `X` and `y` before in hopes that it'd be a bit cleaner, but at least for pipelines/components/other methods, it wasn't always the case that we had both inputs (X and y), and it seemed more clear to me that each input was explicitly transformed separately and then their types converted via `_convert_woodwork_types_wrapper` in a parallel manner right after, so I decided to keep them separate. What are your thoughts on this?

Reply:
@angela97lin I see what you mean! I think it's fine to keep the pattern we have where we call `_convert_to_woodwork_structure` and `_convert_woodwork_types_wrapper` for `X` and `y` separately. I don't think it's too verbose and it's explicit about what's happening. 

What I was suggesting is converting the wrapper you have now into a helper that guarantees that both `X` and `y` are either ww types or None. That's basically what your wrapper is doing now and I think we could use it for components, pipelines, any method. As a case study, here's what `ComponentBase.fit` looks like now:

```python
def fit(self, X, y=None):
    X = _convert_to_woodwork_structure(X)
    X = _convert_woodwork_types_wrapper(X.to_dataframe())
    if y is not None:
        y = _convert_to_woodwork_structure(y)
        y = _convert_woodwork_types_wrapper(y.to_series())
    try:
        self._component_obj.fit(X, y)
        return self
    except AttributeError:
        ....
```
With the "conversion wrapper" it'd be:

```python
def fit(self, X, y=None):
    X, y = convert_to_woodwork(X, y)
    X = _convert_woodwork_types_wrapper(X.to_dataframe())
    if y is not None:
        y = _convert_woodwork_types_wrapper(y.to_series())
    try:
        self._component_obj.fit(X, y)
        return self
    except AttributeError:
```

It's very similar but I think it makes our conventions clearer: When data comes into evalml it should be converted to woodwork. Whether it gets turned back to pandas is up to the logic of the component/pipeline/function.

Let me know what you think! Like I said, I like the pattern we had before and we don't have to let this discussion hold up the merge.

Reply:
@freddyaboulton Ah, I think I see what you mean. I think for now, it'll be okay to keep as is. There are times where we don't necessarily use/need X/y, and this could make it a little clearer when something is ignored. If it gets unwieldy later we can revisit and refactor. This discussion was great though, thanks a lot 😁